skylighting-core/src/Skylighting/Regex.hs

```diff
@@ -2,37 +2,96 @@
 {-# LANGUAGE DeriveDataTypeable  #-}
 {-# LANGUAGE DeriveGeneric       #-}
 {-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE PatternSynonyms #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
 module Skylighting.Regex (
                 Regex(..)
-              , RE(..)
+              , RE
+              , pattern RE, reCaseSensitive, reString
+              , compileRE
               , compileRegex
               , matchRegex
               , testRegex
               , isWordChar
               ) where
 
 import Data.Aeson
-import Data.Binary (Binary)
+import Data.Binary (Binary(..))
 import qualified Data.ByteString.Base64 as Base64
 import qualified Data.ByteString.Char8 as BS
 import Data.Data
 import qualified Data.Text as Text
 import qualified Data.Text.Encoding as TE
-import GHC.Generics (Generic)
 #if !MIN_VERSION_base(4,13,0)
 import Control.Monad.Fail (MonadFail)
 #endif
 import Regex.KDE
 
+import Text.Read hiding (get)
+
 -- | A representation of a regular expression.
-data RE = RE{
-    reString        :: BS.ByteString
-  , reCaseSensitive :: Bool
-} deriving (Show, Read, Ord, Eq, Data, Typeable, Generic)
+data RE = RE'{
+    _reString        :: BS.ByteString
+  , _reCaseSensitive :: Bool
+  , reCompiled'      :: Either String Regex
+} deriving Typeable
+
+-- We define a smart constructor which also holds the compiled regex, to avoid
+-- recompiling each time we tokenize.
+
+{-# COMPLETE RE #-}
+pattern RE :: BS.ByteString -> Bool  -> RE
+pattern RE {reString, reCaseSensitive} <- RE' reString reCaseSensitive _ where
+  RE str caseSensitive = RE' str caseSensitive (compileRegex caseSensitive str)
+
+-- Unfortunately this means we need to derive all the instances ourselves.
+
+instance Show RE where
+  showsPrec d (RE str caseSensitive) = showParen (d > 10) 
+    $ showString "RE {reString = " 
+    . showsPrec 11 str
+    . showString ", reCaseSensitive = "
+    . showsPrec 11 caseSensitive
+    . showString "}"
+
+instance Read RE where
+  readPrec = parens . prec 10 $ do
+    Ident "RE" <- lexP
+    Punc "{" <- lexP
+    Ident "reString" <- lexP
+    Punc "=" <- lexP
+    str <- readPrec
+    Punc "," <- lexP
+    Ident "reCaseSensitive" <- lexP
+    Punc "=" <- lexP
+    caseSensitive <- readPrec
+    Punc "}" <- lexP
+    pure (RE str caseSensitive)
 
```

Comment:
Do I understand correctly that with this change, we'll need to use `RE{ reString = x, reCaseSensitive = y }` instead of `RE x y`?  I don't think that's a big problem, but it should be documented in the release notes.

Reply:
Both syntaxes should be supported - my understanding of the pattern synonym `pattern RE {reString, reCaseSensitive}` is that it's treated the same as a constructor record declaration.

Reply:
Even in the Read instance?  I have not manually written these before.

Reply:
Ahh, apologies. Misunderstood what you were saying there. 

No, the read instance won't support both. Will have a look to see how to add that back.